compat/compat_shared.h

```diff
@@ -248,7 +248,17 @@ const char * squid_strnstr(const char *s, const char *find, size_t slen);
 #if !defined(PRINTF_FORMAT_ARG3)
 #define PRINTF_FORMAT_ARG3 __attribute__ ((format (printf, 3, 4)))
 #endif
-#else /* !__GNU__ */
+#elif __GNUC__
+#if !defined(PRINTF_FORMAT_ARG1)
+#define PRINTF_FORMAT_ARG1 __attribute__ ((format (gnu_printf, 1, 2)))
+#endif
+#if !defined(PRINTF_FORMAT_ARG2)
+#define PRINTF_FORMAT_ARG2 __attribute__ ((format (gnu_printf, 2, 3)))
+#endif
+#if !defined(PRINTF_FORMAT_ARG3)
+#define PRINTF_FORMAT_ARG3 __attribute__ ((format (gnu_printf, 3, 4)))
```

Comment:
> [GCC](https://gcc.gnu.org/onlinedocs/gcc/Common-Function-Attributes.html#index-format-function-attribute): The parameter archetype determines how the format string is interpreted, and should be printf, ... or gnu_printf. On MinGW targets, ms_printf [is] also present.
>
> * archetype values such as printf refer to the formats accepted by the system’s C runtime library, 
> * values prefixed with ‘gnu_’ always refer to the formats accepted by the GNU C Library.
> * On Microsoft Windows targets, values prefixed with ‘ms_’ refer to the formats accepted by the msvcrt.dll library

> PR description: Some build profiles of GCC require the gnu_printf format attribute

If I am interpreting the implications of this PR description statement correctly, some GCC builds warn about some printf() %code used by the official code. That means that %code is not "accepted by the system’s C runtime library". If we enable gnu_printf, we will silence the warning because the offending %code is "accepted by the GNU C Library". However, that warning is correct if the build is _not_ actually using "the GNU C Library". In those cases, we should not silence the warning. Instead, we should change build parameters and/or Squid code.

To avoid such false positives, we can use gnu_printf _only_ when the built is using "the GNU C Library". The proposed `!__clang__ && __GNUC__` condition is different -- as this PR illustrates, `__GNUC__` does not even mean "built using GCC". Can we adjust that condition to match "built using GCC and the GNU C Library"?

Also, _if_ this change is triggered by Windows builds that use msvcrt.dll, then it sounds like we should enable `ms_printf` (for those GCC builds) instead.

In summary, we should use the format attribute that matches the actual build (rather than the format attribute that silences warnings): If a GCC build uses msvcrt.dll, we should use `ms_printf` attribute. If a GCC build uses GNU C Library, we should use `gnu_printf` attribute. Can we adjust these defines to match the current build?

Reply:
That GCC documentation does not tell the whole picture. The archetypes have changed over time as what was "standard" changed. The exact profile for "printf" depends both on which ```-std=``` option is specified, and on macros which can be explicitly set.

Unfortunately MinGW sets a macro to downgrades the "printf" support to the ANSI compliant C89 standard set. Not a C++ compliant set of %codes. The "gnu_printf" archetype is always a superset of the ```-std=``` support, so we can safely use that in its place to get ```%zu``` etc from C++11 and later.

We cannot use the "ms_printf" archetype because that only defines the minimal ANSI set along with a bunch of weird ```%I32...``` and ```%I64...``` prefixed codes for MSVC we should not be polluting Squid with.

Reply:
> MinGW sets a macro to downgrades the "printf" support to the ANSI compliant C89 standard set

I assume the above statement is about what is deemed supported by the "printf" archetype, not what is supported by the actual printf() implementation in that specific MinGW environment.

> The "gnu_printf" archetype is always a superset of the `-std=` support, so we can safely use that in its place to get `%zu` etc from C++11 and later.

AFAICT, these archetypes do not control what is supported by printf(). They control what the compiler warns about when printf() is called.

Given that assumption, I do not understand the "we can safely use a larger set" logic: We are not trying to find the archetype that includes our %codes. We are trying to find the archetype that matches what is actually _supported_ by the current build environment! Obviously, larger archetype sets will produce fewer warnings, but our goal is not fewer warnings. Our goal is detecting unsupported %codes (i.e. correct warnings). Fewer (wrong) warnings is just a side effect.

It is, of course, possible that there is just no archetype that matches our needs and we have to use gnu_printf, even though we risk missing silencing some correct warnings.

> Unfortunately MinGW sets a macro to downgrades the "printf" support

Can we make `gnu_printf` use specific to MinGW then, to reduce the impact of this problem on other `__GNUC__` builds where this "downgrade" of `printf` archetype does not happen?

Reply:
Done in [9762652](https://github.com/squid-cache/squid/pull/1390/commits/97626527cffaa9343c139f37ab7e1e8417fbadea) now that we have the ```os/mingw.h``` header to place the hack.